### PR TITLE
AP-4082: Enable sidekiq dynamic queue naming in UAT

### PR DESCRIPTION
**Pull request by kmahern**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4082)

Describe what you did and why.

- Amended deployment-worker to add a separate container to process each submission "use case" type, plus one for default.
- Added queue_name_service.rb, to define the name of the submissions queue in uat.

This was tested by shelling onto the uat pod, opening a rails console, and entering the following:-

```
TestWorker.perform_async
```

It should then be possible to access the sidekiq console and see that a job has been added to a queue with the branch's name.

Screenshot 2023-05-10 at 13 31 24
Screenshot 2023-05-10 at 13 31 45

## Checklist

Before you ask people to review this PR:

- Tests and linters should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: .helm/assure-hmrc-data/templates/deployment-worker.yaml [hunk 1]
```diff
@@ -48,5 +48,5 @@ spec:
           lifecycle:
             preStop:
               exec:
-                command: ['bundle', 'exec', 'sidekiqctl', 'quiet']
+                command: ['bundle', 'exec', 'sidekiq', '-q', '{{ .Values.sidekiq.queue_name }}submissions', '-c', '1'] # queue_name is set in bin/uat_deploy
```
Comment (kmahern) [inline .helm/assure-hmrc-data/templates/deployment-worker.yaml:51]:
```suggestion
                command: ['bundle', 'exec', 'sidekiq', '-q', '{{ .Values.sidekiq.queue_name }}submissions', '-c', '1'] 
```

File: .env.sample [hunk 2]
```diff
@@ -29,3 +29,5 @@ SIDEKIQ_WEB_UI_PASSWORD='password'
 HMRC_INTERFACE_HOST='https://fake-laa-hmrc-interface-env'
 HMRC_INTERFACE_UID='TestHMRCInterfaceUID'
 HMRC_INTERFACE_SECRET='TestHMRCInterfaceSecret'
+
+ENVIRONMENT='environment'
```
Comment (jsugarman) [inline .env.sample:33]: more idiomatic to use `ENV` or `HOST_ENV`  (inline with RAILS_ENV, RACK_ENV - we did this in CCCD. alternatively could use `HostEnv` to programmatically determine this, as in Apply or `RailsHost` as in CCCD (combinded with `ENV` envvar, for example.

File: .helm/assure-hmrc-data/templates/_envs.tpl [hunk 3]
```diff
@@ -136,4 +136,6 @@ env:
       secretKeyRef:
         name: assure-hmrc-data-application-output
         key: sidekiq_web_ui_password
+  - name: ENVIRONMENT
+    value: {{ .Values.environment | quote }}
```
Comment (jsugarman) [inline .helm/assure-hmrc-data/templates/_envs.tpl:140]: see [comment](https://github.com/ministryofjustice/laa-assure-hmrc-data/pull/177/files#r1189850757)

File: .helm/assure-hmrc-data/values/uat.yaml [hunk 4]
```diff
@@ -49,3 +49,8 @@ mock_azure:
 
 worker:
   replicaCount: 1
+
+environment: uat
```
Comment (jsugarman) [inline .helm/assure-hmrc-data/values/uat.yaml:53]: see [comment](https://github.com/ministryofjustice/laa-assure-hmrc-data/pull/177/files#r1189850757)

File: app/services/queue_name_service.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    new.call
+  end
+
+  def call
+    if Rails.configuration.x.environment == "uat"
+      "#{uat_queue_name}-submissions"
+    else
+      "submissions"
+    end
+  end
```
Comment (jsugarman) [inline app/services/queue_name_service.rb:12]: should now be 

```
 uc-one{-possible-branch-name-onr-nil}-submissions # for all environments
```
could perhaps add (use_case [:one,  :two]) as argument to call

```
  def call(use_case)
    " uc-#{use_case}{#branch_name}-submissions"
  end
  
private
   def branch_name
    return unless if Rails.configuration.x.environment == "uat"
    Rails.configuration.x.status.app_branch.downcase.tr(" _/[]().", "-").prepend("-")
  end
```

```
QueueNameService.call(use_case)
```

Comment (kmahern) [reply app/services/queue_name_service.rb]: 👍 Have made the change to the queue_name_service to reflect the new queue names and added use_case to call as sugggested.

File: .helm/assure-hmrc-data/values/staging.yaml [hunk 6]
```diff
@@ -61,4 +61,4 @@ worker:
 sidekiq:
   queue_name: null
 
-environment: staging
+host_environment: staging
```
Comment (jsugarman) [inline .helm/assure-hmrc-data/values/staging.yaml:64]: host_env

File: .helm/assure-hmrc-data/templates/_envs.tpl [hunk 7]
```diff
@@ -136,4 +136,6 @@ env:
       secretKeyRef:
         name: assure-hmrc-data-application-output
         key: sidekiq_web_ui_password
+  - name: ENVIRONMENT
```
Comment (jsugarman) [inline .helm/assure-hmrc-data/templates/_envs.tpl:139]: I would be consistent

```suggestion
  - name: HOST_ENV
```

File: spec/services/queue_name_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      it "prefixes the submission queue name with the branch name" do
+        expect(call).to eq "uc-one-this-is-a-test-branch-submissions"
+      end
+    end
+
+    context "when the service is called anywhere other than UAT for use case two" do
+      let(:use_case) { 'two' }
+
+      before do
+        Rails.configuration.x.host_env = "test"
```
Comment (jsugarman) [inline spec/services/queue_name_service_spec.rb:24]: see [comment](https://github.com/ministryofjustice/laa-assure-hmrc-data/pull/177/files#r1190741462)

File: spec/services/queue_name_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+RSpec.describe QueueNameService do
+  let(:use_case) { 'one' }
+
+  describe ".call" do
+    subject(:call) { described_class.call(use_case) }
+
+    context "when the service is called in UAT for use case one" do
+      before do
+        Rails.configuration.x.host_env = "uat"
+        Rails.configuration.x.status.app_branch = "this-is/a.test-branch"
```
[the remark this answers is not in the document]
Comment (kmahern) [reply spec/services/queue_name_service_spec.rb]: 👍 have amended but tweaked suggestion slightly as rubocop didn't like message_chain (this cop https://www.rubydoc.info/gems/rubocop-rspec/1.7.0/RuboCop/Cop/RSpec/MessageChain)

File: spec/services/queue_name_service_spec.rb (hunk 8 above)
Comment (kmahern) [reply spec/services/queue_name_service_spec.rb]: 👍 have amended but tweaked suggestion slightly as rubocop didn't like message_chain (this cop https://www.rubydoc.info/gems/rubocop-rspec/1.7.0/RuboCop/Cop/RSpec/MessageChain)

**Review by jsugarman — APPROVED** [review]

**Pull request merged**